README.md
# ReGrid

[![npm version](https://img.shields.io/npm/v/rethinkdb-regrid.svg)](https://www.npmjs.com/package/rethinkdb-regrid) [![license](https://img.shields.io/npm/l/rethinkdb-regrid.svg)](https://github.com/internalfx/regrid/blob/master/LICENSE) [![Build Status](https://travis-ci.org/internalfx/regrid.svg?branch=master)](https://travis-ci.org/internalfx/regrid)

ReGrid is a method of storing large files inside a RethinkDB database.

### Features

- **Reliable** - Files are replicated across the cluster, benefiting from RethinkDB's automatic failover.
- **Scalable** - Easily store large files in RethinkDB, distributed across the cluster.
- **Consistent** - Sha256 hashes are calculated when the file is written, and verified when read back out.
- **Realtime** - Watch the file system for changes and be notified immediately.
- **Fast** - Supports in-memory caching so your server doesn't fall over when something goes viral.
- **Current** - Uses RethinkDB's changefeeds to evict cached items. ReGrid doesn't serve stale files.

View the [Changelog](https://github.com/internalfx/regrid/blob/master/CHANGELOG.md)

The [ReGrid spec](https://github.com/internalfx/regrid-spec) is an open specification free for anyone to implement and use.

**The spec is very out of date**

#### Need help?

Join Slack [here](http://slack.rethinkdb.com/), then meet us on the [#regrid](https://rethinkdb.slack.com/messages/regrid/) channel.

[![Join Slack](http://slack.rethinkdb.com/badge.svg)](http://slack.rethinkdb.com/)

---

Special thanks to [Arthur Andrew Medical](http://www.arthurandrew.com/) for sponsoring this project.

Arthur Andrew Medical manufactures products with ingredients that have extensive clinical research for safety and efficacy. We specialize in Enzymes, Probiotics and Antioxidants.

---

## Installation

Supports node v8.0+

```
npm install --save rethinkdb-regrid
```

## TL;DR

```javascript
var ReGrid = require('rethinkdb-regrid')

var bucket = ReGrid({db: 'example'})

// initBucket creates tables and indexes if they don't exist, returns a promise.
bucket.initBucket().then(function () {
  // We are now ready to read and write files

  // Watch a filename for changes
  bucket.watchFilename('/videos/myVid.mp4').then(function (cursor) {
    cursor.each((err, changes) => {
      console.log(changes)
    })
  })

  // Open a write stream to ReGrid
  var dbStream = bucket.upload('/videos/bigvid.mp4')

  // Create read stream from file and pipe it to a ReGrid write stream
  fs.createReadStream('./bigvid.mp4').pipe(dbStream)

  dbStream.on('finish', function () {
    // File is now written to the database

    //Read the file and pipe it to a write stream to save the file back out to the file system.
    bucket.downloadFilename('/videos/bigvid.mp4').pipe(fs.createWriteStream('./copy-of-bigvid.mp4'))
  })

})
```

## API Documentation

### `ReGrid([connectionOptions, options])`

##### Parameters

| key | default | type | description |
| --- | --- | --- | --- |
| connectionOptions | {}| Object | `connectionOptions` is passed directly to [rethinkdbdash](https://github.com/neumino/rethinkdbdash) |
| options | {} | Object |  Optional parameters listed below |

###### Options

| key | default | type | description |
|---|---|---|---|
| bucketName | `fs` | String | The name of the bucket. Table names are prefixed by this. |
| chunkSizeBytes | `1024 * 255` | Number | The default chunk size, in bytes. |
| concurrency | `10` | Number | When writing/reading a file, the number of concurrent queries in flight for a given stream. |
| cacheSize | `400` | Number | The maximum number of objects to keep in memory. Setting to `0` will disable caching. |

##### returns

`Bucket instance`

##### Description

Creates a new ReGrid bucket instance.

##### Example

```javascript
var ReGrid = require('rethinkdb-regrid')

var bucket = ReGrid({db: 'mydatabase'}, {bucketName: 'mybucket'})
```

---

### `initBucket()`

##### Parameters

none

##### returns

Promise

##### Description

Verifies required tables and indexes exist and will create them if missing.

##### Example

```javascript
bucket.initBucket().then(function () {
  // bucket is ready for use.....
})
```

---

### `writeFile(options)`

###### Options

| key | default | type | description |
| --- | --- | --- | --- |
| filename | *required* | String | The name of the file. |
| buffer | *required* | Buffer | A buffer of file contents. |
| chunkSizeBytes | The `chunkSizeBytes` setting for the bucket. | Number | Size of each chunk in bytes. |
| metadata | {} | Object | Metadata object allowing you to store custom keys on files. |

##### returns

Promise

##### Description

Returns a promise that resolves to the newly written file.

##### Example

```javascript
let fileBuffer = fs.readFileSync('./myVid.mp4')

let newFile = await bucket.writeFile({filename: '/videos/myVid.mp4', buffer: fileBuffer})
```

---

### `createWriteStream(options)`

###### Options

| key | default | type | description |
| --- | --- | --- | --- |
| filename | *required* | String | The name of the file. |
| chunkSizeBytes | The `chunkSizeBytes` setting for the bucket. | Number | Size of each chunk in bytes. |
| metadata | {} | Object | Metadata object allowing you to store custom keys on files. |

##### returns

WriteStream

##### Description

Returns a write stream for storing a file in ReGrid.

##### Example

```javascript
var writeStream = bucket.createWriteStream({
  filename: '/videos/myVid.mp4',
  chunkSizeBytes: 1024 * 255,
  metadata: {topic: 'cats'}
})

writeStream.on('finish', function () {
  // File is now stored in ReGrid
})

fs.createReadStream('./myVid.mp4').pipe(writeStream)
```

---

### `getFile(options)`

###### Options

| key | default | type | description |
| --- | --- | --- | --- |
| id | Null | String | The `id` of the file to retrieve. |
| filename | Null | String | Ignored if `id != null`. The `filename` of the file to retrieve |
| revision | `-1` | Number | Ignored if `id != null`. The revision of the file to retrieve. If multiple files are uploaded under the same `filename` they are considered revisions. This may be a positive or negative number. (see chart below) |

###### How revision numbers work

If there are five versions of a file, the below chart would be the revision numbers

| Number | Description |
| --- | --- |
| `0` or `-5` | The original file |
| `1` or `-4` | The first revision |
| `2` or `-3` | The second revision |
| `3` or `-2` | The second most recent revision |
| `4` or `-1` | The most recent revision |

##### Description

Returns a promise that resolves to the files information.

##### Example

```javascript
let file1 = bucket.getFile({id: 'ca608825-15c0-44b5-9bef-3ccabf061bab'})
let file2 = bucket.getFile({filename: 'catVideo.mp4', revision: 2})
```

---

### `readFile(options)`

###### Options

| key | default | type | description |
| --- | --- | --- | --- |
| id | Null | String | The `id` of the file to retrieve. |
| filename | Null | String | Ignored if `id != null`. The `filename` of the file to retrieve |
| revision | `-1` | Number | Ignored if `id != null`. The revision of the file to retrieve. If multiple files are uploaded under the same `filename` they are considered revisions. This may be a positive or negative number. (see chart below) |
| seekStart | Null | Number | The start of the byte range. |
| seekEnd | Null | Number | The end of the byte range. If omitted the stream will continue to the end of file. |

###### How revision numbers work

If there are five versions of a file, the below chart would be the revision numbers

| Number | Description |
| --- | --- |
| `0` or `-5` | The original file |
| `1` or `-4` | The first revision |
| `2` or `-3` | The second revision |
| `3` or `-2` | The second most recent revision |
| `4` or `-1` | The most recent revision |

##### Description

Returns a promise that resolves to the files information and contents.

##### Example

```javascript
let file1 = bucket.readFile({id: 'ca608825-15c0-44b5-9bef-3ccabf061bab'})
let file2 = bucket.readFile({filename: 'catVideo.mp4', revision: 2})
```

---

### `createReadStream(options)`

###### Options

| key | default | type | description |
| --- | --- | --- | --- |
| id | *required* | String | The `id` of the file to retrieve |
| seekStart | Null | Number | The start of the byte range. |
| seekEnd | Null | Number | The end of the byte range. If omitted the stream will continue to the end of file. |

##### returns

ReadStream

##### Description

Returns a read stream for reading a file from ReGrid.

##### Example

```javascript
var readStream = bucket.createReadStream({id: 'ca608825-15c0-44b5-9bef-3ccabf061bab'})

readStream.pipe(fs.createWriteStream('./mySavedVideo.mp4'))
```

---

### `downloadFilename(filename[, options])`

##### Parameters

| key | default | type | description |
| --- | --- | --- | --- |
| filename | *required* | String | The name of the file. |
| options | {} | Object |  Optional parameters listed below |

###### Options

| key | default | type | description |
| --- | --- | --- | --- |
| revision | `-1` | Number | The revision of the file to retrieve. If multiple files are uploaded under the same `filename` they are considered revisions. This may be a positive or negative number. (see chart below) |
| seekStart | Null | Number | The start of the byte range. |
| seekEnd | Null | Number | The end of the byte range. If omitted the stream will continue to the end of file. |

###### How revision numbers work

If there are five versions of a file, the below chart would be the revision numbers

| Number | Description |
| --- | --- |
| `0` or `-5` | The original file |
| `1` or `-4` | The first revision |
| `2` or `-3` | The second revision |
| `3` or `-2` | The second most recent revision |
| `4` or `-1` | The most recent revision |

##### returns

ReadStream

##### Description

Returns a read stream for reading a file from ReGrid.

##### Example

```javascript
var newestVersion = bucket.downloadFilename('/videos/myVid.mp4')

var originalVersion = bucket.downloadFilename('/videos/myVid.mp4', {revision: 0})

newestVersion.pipe(fs.createWriteStream('./latest.mp4'))

originalVersion.pipe(fs.createWriteStream('./original.mp4'))
```

---

### `getFilename(filename[, options])`

##### Parameters

| key | default | type | description |
| --- | --- | --- | --- |
| filename | *required* | String | The name of the file. |
| options | {} | Object |  Optional parameters listed below |

###### Options

| key | default | type | description |
| --- | --- | --- | --- |
| revision | `-1` | Number | The revision of the file to retrieve. If multiple files are uploaded under the same `filename` they are considered revisions. This may be a positive or negative number. (see chart below) |

###### How revision numbers work

If there are five versions of a file, the below chart would be the revision numbers

| Number | Description |
| --- | --- |
| `0` or `-5` | The original file |
| `1` or `-4` | The first revision |
| `2` or `-3` | The second revision |
| `3` or `-2` | The second most recent revision |
| `4` or `-1` | The most recent revision |

##### returns

Promise that resolves to the files information.

##### Description

Returns information for a file from ReGrid.

##### Example

```javascript
bucket.getFilename('/videos/myVid.mp4', {revision: 0}).then(function (file) {
  console.log(file)

  /*
  RETURNS
  {
    "chunkSizeBytes": 261120 ,
    "filename":  "/videos/myVid.mp4" ,
    "finishedAt": Wed Jan 11 2017 20:35:20 GMT+00:00 ,
    "id":  "2578d295-6041-40e6-9235-d02e96c524ca" ,
    "length": 174110 ,
    "sha256":  "298fac8be0bdb1e48fa520dee5f510ef5f91ea0411b7584e8c62ec0cf34ce932" ,
    "startedAt": Wed Jan 11 2017 20:35:20 GMT+00:00 ,
    "status":  "Complete"
  }
  */
})
```

---

### `listRegex(pattern[, options])`

##### Parameters

| key | default | type | description |
| --- | --- | --- | --- |
| pattern | *required* | String | A regular expression matched against `filename` |
| options | {} | Object |  Optional parameters listed below |

###### Options

| key | default | type | description |
| --- | --- | --- | --- |
| sort | 'DESC' | String | Sort results by `filename`. Valid values are `ASC` and `DESC`. |
| skip | undefined | Number | Skip results, useful for pagination. |
| limit | undefined | Number | Limit results. |
| showAll | false | Boolean | Show all revisions of matched files. Normally duplicate filenames are hidden and only the latest revision of each filename is returned. |

##### returns

ReadStream in `objectMode`. The stream emits objects, not buffers.

##### Description

Returns a read stream for finding files via regular expression. The stream emits objects as they are found, and can be a long running operation.

You may call `toArray()` on the returned stream to coerce it to an array. This may fail if the result set is very large. You may optionally call with limit to prevent this.

##### Example

```javascript
var videoStream = bucket.listRegex('^/videos/', {limit: 100})

videoStream.on('data', function (video) {
  console.log(video)
})

// OR
bucket.listRegex('^/videos/', {limit: 100}).toArray().then(function (videos) {
  // list the first 100 videos in the `/videos` directory
  console.log(videos)
})
```

---

### `listFilename(filename[, options])`

##### Parameters

| key | default | type | description |
| --- | --- | --- | --- |
| filename | *required* | String | The name of the file. |
| options | {} | Object |  Optional parameters listed below |

###### Options

| key | default | type | description |
| --- | --- | --- | --- |
| sort | undefined | String | Sort results by `finishedAt` (The date the file was uploaded). Valid values are `ASC` and `DESC`. |
| skip | undefined | Number | Skip results, useful for pagination. |
| limit | undefined | Number | Limit results. |

##### returns

ReadStream in `objectMode`. The stream emits objects, not buffers.

##### Description

Returns a read stream for finding files by filename. The stream emits objects as they are found, and can be a long running operation. However, this method uses an index and is very efficient.

You may call `toArray()` on the returned stream to coerce it to an array. This may fail if the result set is very large. You may optionally call with limit to prevent this.

##### Example

```javascript
var revisionStream = bucket.listFilename('/videos/editedVideo.mp4')

revisionStream.on('data', function (revision) {
  console.log(revision)
})

// OR
bucket.listFilename('/videos/editedVideo.mp4').toArray().then(function (revisions) {
  // list all revisions of '/videos/editedVideo.mp4'
  console.log(revisions)
})
```

---

### `listMetadata(metadata[, options])`

##### Parameters

| key | default | type | description |
| --- | --- | --- | --- |
| metadata | *required* | Object | An object to match against the `metadata` field of ReGrid files. |
| options | {} | Object |  Optional parameters listed below |

###### Options

| key | default | type | description |
| --- | --- | --- | --- |
| skip | undefined | Number | Skip results, useful for pagination. |
| limit | undefined | Number | Limit results. |

##### returns

ReadStream in `objectMode`. The stream emits objects, not buffers.

##### Description

Returns a read stream for finding files by metadata. The stream emits objects as they are found, and can be a long running operation.

You may call `toArray()` on the returned stream to coerce it to an array. This may fail if the result set is very large. You may optionally call with limit to prevent this.

##### Example

```javascript
var catStream = bucket.listMetadata({topic: 'cats'})

catStream.on('data', function (catVideo) {
  console.log(catVideo)
})

// OR
bucket.listMetadata({topic: 'cats'}).toArray().then(function (catVideos) {
  // list all videos under the 'cat' topic.
  console.log(catVideos)
})
```

---

### `watchRegex(pattern)`

##### Parameters

| key | default | type | description |
| --- | --- | --- | --- |
| pattern | *required* | String | A regular expression matched against filename. |

##### returns

A RethinkDB [changeFeed](https://www.rethinkdb.com/api/javascript/changes/) cursor

##### Description

Allows you to be notified of changes to a file if the `filename` property matches your regular expression.

##### Example

```javascript
bucket.watchRegex('^/videos').then(function (cursor) {
  cursor.each((err, video) => {
    console.log(video)
  })
})
```

---

### `watchFilename(filename)`

##### Parameters

| key | default | type | description |
| --- | --- | --- | --- |
| filename | *required* | String | The `filename` to watch for changes. |

##### returns

A RethinkDB [changeFeed](https://www.rethinkdb.com/api/javascript/changes/) cursor

##### Description

Allows you to be notified of changes to a file if the `filename` property `===` your string.

##### Example

```javascript
bucket.watchFilename('/videos/myVid.mp4').then(function (cursor) {
  cursor.each((err, changes) => {
    console.log(changes)
  })
})
```

---

### `watchMetadata(metadata)`

##### Parameters

| key | default | type | description |
| --- | --- | --- | --- |
| metadata | *required* | Object | The `metadata` to watch for changes. |

##### returns

A RethinkDB [changeFeed](https://www.rethinkdb.com/api/javascript/changes/) cursor

##### Description

Allows you to be notified of changes to a file if the `metadata` property matches your object.

##### Example

```javascript
bucket.watchMetadata({topic: 'cats'}).then(function (cursor) {
  cursor.each((err, catVideoChanges) => {
    console.log(catVideoChanges)
  })
})
```

---

### `deleteId(fileId)`

##### Parameters

| key | default | type | description |
| --- | --- | --- | --- |
| fileId | *required* | String | The `id` of the file to delete |

##### returns


CHANGELOG.md
# 0.8.1

- Bug fixes

# 0.8.0

#### New Features

- Added `writeFile` function
- Added `createWriteStream` function
- Added `getFile` function
- Added `readFile` function
- Added `createReadStream` function

#### Breaking Changes

- Removed `upload` function
- Removed `downloadId` function
- Removed `downloadFilename` function
- Removed `getFilename` function

# 0.7.0

#### New Features

- Added `getFilename` function
- Added in-memory caching for files and chunks

#### Breaking Changes

- Renamed `rename` function to `renameId`
- Renamed `delete` function to `deleteId`


lib/regrid.js
'use strict'

const Promise = require('bluebird')
Promise.config({
  warnings: {
    wForgottenReturn: false
  }
})

const rgUtils = require('./RGUtils')
const Readable = require('stream').Readable
const Writable = require('stream').Writable
const crypto = require('crypto')
const toArray = require('stream-to-array')
const lruCache = require('lru-cache')

let includes = function (list, item) {
  return list.indexOf(item) !== -1
}

const defaultBucketOptions = {
  bucketName: 'fs',
  chunkSizeBytes: 1024 * 255,
  concurrency: 10,
  cacheSize: 400
}

let ReGrid = function (connOpts, bucketOptions) {
  let conf = Object.assign({}, defaultBucketOptions, bucketOptions)
  let fileTable = `${conf.bucketName}_files`
  let chunkTable = `${conf.bucketName}_chunks`
  let r = require('rethinkdbdash')(Object.assign({silent: true}, connOpts))
  let isCaching = conf.cacheSize > 0
  let cache

  let getFileKey = function (filename) {
    return `file-${filename}`
  }

  let fetchFileById = async function (id) {
    let fileKey = `file-${id}`
    let file

    if (isCaching) {
      file = cache.get(fileKey)
    }

    if (file == null) {
      file = await r.table(fileTable).get(id).default(null).run()
      if (file != null && isCaching) {
        cache.set(fileKey, file)
      }
    }

    return file
  }

  let fetchFileByName = async function (filename, revision = -1) {
    let fileKey = getFileKey(filename)
    let file

    if (revision === -1 && isCaching) {
      file = cache.get(fileKey)
    }

    if (file == null) {
      let revSteps
      let query = r.table(fileTable).between(['Complete', filename, r.minval], ['Complete', filename, r.maxval], {index: 'file_ix'})
      if (revision >= 0) {
        revSteps = revision
        query = query.orderBy({index: r.asc('file_ix')})
      } else {
        revSteps = (revision * -1) - 1
        query = query.orderBy({index: r.desc('file_ix')})
      }

      let files = await query.run()

      file = files[revSteps] || null

      if (file != null && revision === -1 && isCaching) {
        cache.set(fileKey, file)
      }
    }

    return file
  }

  let createChunkLoader = function (file, startNum, endNum) {
    let num = startNum
    let queueNum = startNum
    let queue = []

    let next = async function () {
      if (num > endNum) { return }

      while (queue.length < conf.concurrency && queueNum <= endNum) {
        let chunkKey = `chunk-${file.id}-${queueNum}`
        let chunk

        if (isCaching) {
          chunk = cache.get(chunkKey)
        }

        if (chunk != null) {
          queue.push({key: chunkKey, chunk: Promise.resolve(chunk), cached: true})
        } else {
          queue.push({key: chunkKey, chunk: r.table(chunkTable).getAll([file.id, queueNum], {index: 'chunk_ix'}).nth(0).default(null).run(), cached: false})
        }
        queueNum += 1
      }

      let chunkData = queue.shift()
      let chunk = await chunkData.chunk
      if (chunkData.cached === false && isCaching) {
        cache.set(chunkData.key, chunk)
      }

      num += 1
      return chunk
    }

    return {
      next
    }
  }

  let initBucket = async function () {
    var promises = []
    var tableList = await r.tableList()

    // Create tables if they don't exist.
    if (!includes(tableList, fileTable)) {
      promises.push(r.tableCreate(fileTable))
    }
    if (!includes(tableList, chunkTable)) {
      promises.push(r.tableCreate(chunkTable))
    }

    await Promise.all(promises)

    var result = await Promise.all([
      r.table(fileTable).indexList(),
      r.table(chunkTable).indexList()
    ])

    promises = []

    var fileIndexList = result[0]
    var chunkIndexList = result[1]

    // Create indexes if they don't exist.
    if (!includes(fileIndexList, 'file_ix')) {
      promises.push(r.table(fileTable).indexCreate('file_ix', [r.row('status'), r.row('filename'), r.row('finishedAt')]))
    }

    if (!includes(chunkIndexList, 'chunk_ix')) {
      promises.push(r.table(chunkTable).indexCreate('chunk_ix', [r.row('file_id'), r.row('num')]))
    }

    await Promise.all(promises)

    await Promise.all([
      r.table(fileTable).indexWait('file_ix'),
      r.table(chunkTable).indexWait('chunk_ix')
    ])

    if (isCaching) {
      cache = lruCache({
        max: conf.cacheSize
      })

      r.table(fileTable).changes().run(function (err, cursor) {
        if (err) { console.log(err) }
        cursor.each(function (err, row) {
          if (err) {
            // console.log(err)
          } else {
            let newVal = row.new_val

            if (newVal) {
              if (newVal.status === 'Complete') {
                cache.del(getFileKey(newVal.filename))
              } else if (newVal.status === 'Deleted') {
                let fileKey = `file-${newVal.filename}`
                cache.del(fileKey)
              }
            }
          }
        })
      })
    }
  }

  let writeFile = async function (spec = {}) {
    spec = Object.assign({
      filename: null,
      buffer: null,
      chunkSizeBytes: conf.chunkSizeBytes,
      metadata: {}
    }, spec)

    if (spec.buffer == null) { throw new Error('buffer must not be null') }

    let wstream = createWriteStream(spec)
    let uploadPromise = rgUtils.writeStreamPromise(wstream)
    wstream.write(spec.buffer)
    wstream.end()
    await uploadPromise

    let file = await fetchFileByName(spec.filename)
    return file
  }

  let createWriteStream = function (spec = {}) {
    spec = Object.assign({
      filename: null,
      chunkSizeBytes: conf.chunkSizeBytes,
      metadata: {}
    }, spec)

    if (spec.filename == null) { throw new Error('filename must not be null') }

    let stream = new Writable()
    let chunkSizeBytes = spec.chunkSizeBytes
    let num = 0
    let currentChunk = Buffer.from([])
    let queue = []
    let hash = crypto.createHash('sha256')
    let size = 0
    let file

    stream._write = async function (chunk, encoding, cb) {
      size += chunk.length
      hash.update(chunk)

      if (file == null) {
        file = await r.table(fileTable).insert({
          filename: spec.filename,
          chunkSizeBytes: chunkSizeBytes,
          startedAt: new Date(),
          status: 'Incomplete',
          metadata: spec.metadata
        }, {returnChanges: 'always'}).run()
        file = file.changes[0].new_val
      }

      currentChunk = Buffer.concat([currentChunk, chunk])

      if (currentChunk.length >= chunkSizeBytes) {
        queue.push(r.table(chunkTable).insert({
          file_id: file.id,
          num: num,
          data: currentChunk.slice(0, chunkSizeBytes)
        }).run())
        currentChunk = currentChunk.slice(chunkSizeBytes)
        num += 1

        if (queue.length >= 10) {
          await Promise.any(queue)
          queue = queue.filter((promise) => promise.isPending())
        }
      }

      cb()
    }

    stream._final = async function (cb) {
      if (file) {
        let chunks = []
        while (currentChunk.length > chunkSizeBytes) {
          chunks.push(currentChunk.slice(0, chunkSizeBytes))
          currentChunk = currentChunk.slice(chunkSizeBytes)
        }
        chunks.push(currentChunk)

        for (let chunk of chunks) {
          queue.push(r.table(chunkTable).insert({
            file_id: file.id,
            num: num,
            data: chunk
          }).run())
          num += 1
        }

        await Promise.all(queue)

        file = await r.table(fileTable).get(file.id).update({
          finishedAt: new Date(),
          length: size,
          status: 'Complete',
          sha256: hash.digest('hex')
        }, {returnChanges: 'always'}).run()
        file = file.changes[0].new_val
      } else {
        // If we are here the file must be empty!
        file = await r.table(fileTable).insert({
          filename: spec.filename,
          finishedAt: new Date(),
          startedAt: new Date(),
          length: size,
          status: 'Complete',
          sha256: hash.digest('hex'),
          chunkSizeBytes: chunkSizeBytes,
          metadata: spec.metadata
        }, {returnChanges: 'always'}).run()
        file = file.changes[0].new_val
      }

      if (isCaching) {
        cache.set(getFileKey(spec.filename), file)
      }

      cb()
    }

    return stream
  }

  let getFile = async function (spec = {}) {
    spec = Object.assign({
      filename: null,
      revision: -1,
      id: null
    }, spec)

    if (spec.id == null && spec.filename == null) { throw new Error('filename or id required') }

    let file

    if (spec.id != null) {
      file = await fetchFileById(spec.id)
    } else {
      file = await fetchFileByName(spec.filename, spec.revision)
    }

    return file
  }

  let readFile = async function (spec = {}) {
    spec = Object.assign({
      seekStart: null,
      seekEnd: null
    }, spec)

    let file = await getFile(spec)

    if (file != null) {
      let stream = createReadStream({id: file.id, seekStart: spec.seekStart, seekEnd: spec.seekEnd})
      file.buffer = await rgUtils.readStreamPromise(stream)
    }

    return file
  }

  let createReadStream = function (spec = {}) {
    spec = Object.assign({
      id: null,
      seekStart: null,
      seekEnd: null
    }, spec)

    if (spec.id == null) { throw new Error('id must not be null') }

    let stream = new Readable()
    stream.toArray = toArray
    let hash = crypto.createHash('sha256')
    let num = 0
    let startNum = 0
    let startOffset = 0
    let endNum = null
    let endOffset = null
    let chunkSize
    let chunkLoader
    let file
    let verifyHash = (spec.seekStart == null && spec.seekEnd == null)
    let isDestroyed = false

    stream._destroy = function () {
      isDestroyed = true
    }

    stream._read = async function () {
      let chunk

      if (file == null) {
        file = await fetchFileById(spec.id)
        chunkSize = file.chunkSizeBytes
        if (spec.seekStart != null) {
          startNum = Math.floor(spec.seekStart / chunkSize)
          startOffset = spec.seekStart % chunkSize
        }

        spec.seekEnd = spec.seekEnd || file.length + 1
        endNum = Math.floor((spec.seekEnd) / chunkSize)
        endOffset = (spec.seekEnd % chunkSize) + 1

        num = startNum
        chunkLoader = createChunkLoader(file, startNum, endNum)
      }

      chunk = await chunkLoader.next()

      if (isDestroyed) { return }

      if (chunk) {
        if (verifyHash) { hash.update(chunk.data) }
        if (num === startNum && num === endNum) {
          stream.push(chunk.data.slice(startOffset, endOffset))
        } else if (num === startNum) {
          stream.push(chunk.data.slice(startOffset))
        } else if (num === endNum) {
          stream.push(chunk.data.slice(0, endOffset))
        } else {
          stream.push(chunk.data)
        }
        num += 1
      } else {
        if (verifyHash) {
          var sha256 = hash.digest('hex')
          if (sha256 !== file.sha256) {
            process.nextTick(function () {
              stream.emit('error', new Error('sha256 hash mismatch: File is likely corrupted!'))
            })
            return
          }
        }
        stream.push(null)
      }
    }

    return stream
  }

  let listRegex = function (pattern, options = {}) {
    options = Object.assign({
      sort: 'DESC',
      showAll: false
    }, options)

    let stream = new Readable({objectMode: true})
    stream.toArray = toArray
    let cursor
    let skipped = 0
    let pushed = 0
    let lastFile = {}

    stream._read = async function () {
      let nextPush

      if (cursor == null) {
        let query = r.table(fileTable).between(['Complete', r.minval], ['Complete', r.maxval], {index: 'file_ix'})

        if (options.sort === 'ASC') {
          query = query.orderBy({index: r.asc('file_ix')})
        } else if (options.sort === 'DESC') {
          query = query.orderBy({index: r.desc('file_ix')})
        }

        query = query.filter(r.row('filename').match(pattern))

        cursor = await query.run({cursor: true})
      }

      while (typeof nextPush === 'undefined') {
        let file

        try {
          file = await cursor.next()
        } catch (err) {
          if (err.message !== 'No more rows in the cursor.') {
            throw err
          }
        }

        if (file) {
          if (options.showAll) {
            nextPush = file
          } else {
            if (file.filename !== lastFile.filename) {
              if (options.sort === 'DESC') {
                nextPush = file
              } else if (options.sort === 'ASC') {
                if (typeof lastFile.filename !== 'undefined') {
                  nextPush = lastFile
                }
              }
            }
          }
          lastFile = file
        } else {
          if (options.sort === 'ASC' && lastFile && lastFile.filename) {
            nextPush = lastFile
            lastFile = null
          } else {
            nextPush = null
          }
        }

        if (typeof nextPush !== 'undefined') {
          if (isFinite(options.skip) && options.skip > skipped) {
            nextPush = undefined
            skipped += 1
          }
        }

        if (typeof nextPush !== 'undefined') {
          if (isFinite(options.limit)) {
            if (options.limit > pushed) {
              pushed += 1
            } else {
              nextPush = null
            }
          }
        }
      }

      stream.push(nextPush)
    }

    return stream
  }

  let listFilename = function (filename, options = {}) {
    let stream = new Readable({objectMode: true})
    stream.toArray = toArray
    let cursor = null

    stream._read = async function () {
      let file

      if (cursor == null) {
        let query = r.table(fileTable).between(['Complete', filename, r.minval], ['Complete', filename, r.maxval], {index: 'file_ix'})

        if (options.sort === 'ASC') {
          query = query.orderBy({index: r.asc('file_ix')})
        } else if (options.sort === 'DESC') {
          query = query.orderBy({index: r.desc('file_ix')})
        }

        if (isFinite(options.skip)) {
          query = query.skip(options.skip)
        }

        if (isFinite(options.limit)) {
          query = query.limit(options.limit)
        }

        cursor = await query.run({cursor: true})
      }

      try {
        file = await cursor.next()
      } catch (err) {
        if (err.message !== 'No more rows in the cursor.') {
          throw err
        }
      }

      if (file) {
        stream.push(file)
      } else {
        stream.push(null)
      }
    }

    return stream
  }

  let listMetadata = function (metadata, options = {}) {
    let stream = new Readable({objectMode: true})
    stream.toArray = toArray
    let cursor = null

    stream._read = async function () {
      let file

      if (cursor == null) {
        let query = r.table(fileTable).filter({metadata: metadata, status: 'Complete'})

        if (isFinite(options.skip)) {
          query = query.skip(options.skip)
        }

        if (isFinite(options.limit)) {
          query = query.limit(options.limit)
        }

        cursor = await query.run({cursor: true})
      }

      try {
        file = await cursor.next()
      } catch (err) {
        if (err.message !== 'No more rows in the cursor.') {
          throw err
        }
      }

      if (file) {
        stream.push(file)
      } else {
        stream.push(null)
      }
    }

    return stream
  }

  let watchRegex = function (pattern) {
    return r.table(fileTable)
    .filter(r.and(r.row('status').eq('Complete'), r.row('filename').match(pattern)))
    .changes()
    .run({cursor: true})
  }

  let watchFilename = function (filename) {
    return r.table(fileTable)
    .between(['Complete', filename, r.minval], ['Complete', filename, r.maxval], {index: 'file_ix'})
    .changes()
    .run({cursor: true})
  }

  let watchMetadata = function (metadata) {
    return r.table(fileTable)
    .filter({metadata: metadata, status: 'Complete'})
    .changes()
    .run({cursor: true})
  }

  let deleteId = async function (fileId, options) {
    options = Object.assign({
      hard: false
    }, options)

    var result = await r.table(fileTable).get(fileId).update({status: 'Deleted', deletedAt: new Date()}).default({}).run()
    if (options.hard === true) {
      await r.table(chunkTable).between([fileId, r.minval], [fileId, r.maxval], {index: 'chunk_ix'}).delete()
      await r.table(fileTable).get(fileId).delete()
    }
    return result.replaced === 1
  }

  let renameId = async function (fileId, filename) {
    var result = await r.table(fileTable).get(fileId).update({filename: filename}).run()
    return result.replaced === 1 || result.unchanged === 1
  }

  let getMetadata = async function (fileId) {
    var result = await r.table(fileTable).get(fileId).run()
    return result
  }

  let replaceMetadata = async function (fileId, metadata) {
    var result = await r.table(fileTable).get(fileId).update({metadata: metadata}).run()
    return result.replaced === 1 || result.unchanged === 1
  }

  return Object.freeze({
    initBucket,
    writeFile,
    createWriteStream,
    getFile,
    readFile,
    createReadStream,
    listRegex,
    listFilename,
    listMetadata,
    watchRegex,
    watchFilename,
    watchMetadata,
    deleteId,
    renameId,
    getMetadata,
    replaceMetadata
  })
}

module.exports = ReGrid
